Add paper for JuliaCon proceedings (#322)
xKDR/Survey.jl: paper/paper.tex

% \href{https://www.hcp.med.harvard.edu/statistics/survey-soft/bascula.html}{Bascula}, 
% \href{https://www.hcp.med.harvard.edu/statistics/survey-soft/cenvar.html}{CENVAR}, 
% \href{https://www.hcp.med.harvard.edu/statistics/survey-soft/clusters.html}{CLUSTERS},  
% \href{https://www.cdc.gov/epiinfo/index.html}{Epi Info},  
% \href{https://www.statcan.gc.ca/eng/survey/methodology/Generalized_Estimation_System-eng.htm}{Generalized Estimation System (GES)},  
% \href{https://isr.umich.edu/}{IVEware},  
% \href{https://catalog.iastate.edu/azcourses/stat/}{PCCARP},  
% \href{https://cran.r-project.org/package=survey}{R survey package}, 
% \href{https://www.sas.com/en_us/home.html}{SAS/STAT},  
% \href{https://www.ibm.com/products/spss-statistics}{SPSS Complex Samples},  
% \href{https://www.stata.com/}{Stata},  
% \href{https://sudaanorder.rti.org/}{SUDAAN},  
% \href{https://www.census.gov/data/software/vplx.html}{VPLX},  
% \href{https://www.westat.com/wesvar/}{WesVar}

% The survey package in R by Thomas Lumely \cite{lumley2004analysis} is the widely used open-source package. 

\section{Survey design}

> A \verb|SurveyDesign| object can be created to incorporate the sampling design. This object requires the following parameters: \verb|data::DataFrame|, which is the survey data in the form of a \verb|DataFrame|; \verb|clusters::Symbol|, specifying the column name containing the clusters; \verb|strata::Symbol|, specifying the column name containing the strata; \verb|weights::Symbol|, indicating the column name containing the sampling weights; and \verb|popsize::Symbol|, indicating the column name containing the population size.\footnote{Internaly, there is a single constructor for all types of surveys. Every survey is assumed to be a complex survey. If there is no stratification, we assume that everything is part of one stratum. If there is no clustering, we assume each member is a cluster.}  

Reviewer: `This object requires the following parameters`: instead of `parameters`, should we say `properties` instead? Usually parameters are constants/values that tag some method/type, or I like to think of them as tunable values associated to some model.

Author: How about `argument`?

Reviewer: `argument` definitely sounds better.

Author: > How about `argument`?

```suggestion
A \verb|SurveyDesign| object can be created to incorporate the sampling design. This object requires the following arguments: \verb|data::DataFrame|, which is the survey data in the form of a \verb|DataFrame|; \verb|clusters::Symbol|, specifying the column name containing the clusters; \verb|strata::Symbol|, specifying the column name containing the strata; \verb|weights::Symbol|, indicating the column name containing the sampling weights; and \verb|popsize::Symbol|, indicating the column name containing the population size.\footnote{Internaly, there is a single constructor for all types of surveys. Every survey is assumed to be a complex survey. If there is no stratification, we assume that everything is part of one stratum. If there is no clustering, we assume each member is a cluster.}  
```

For example, consider the NHANES dataset, which includes clustering and stratification. The following example demonstrates how to create a \verb|SurveyDesign| object for this dataset:
\begin{lstlisting}
julia> nhanes = load_data("nhanes"); 
# CSV dataframe included with the package

julia> design = SurveyDesign(nhanes;
                    clusters=:SDMVPSU,
                    strata=:SDMVSTRA, 
                    weights=:WTMEC2YR);
    \end{lstlisting}
Consider another example, a cluster sample based on the Academic Performance Index for all California schools based on standardised testing of students. There is no stratification in this example. 

\begin{lstlisting}
julia> apiclus1 = load_data("apiclus1"); 
# CSV dataframe included with the package

> design = SurveyDesign(apiclus1; 

Reviewer: `julia> ` missing here?

Author: > `julia> ` missing here?

```suggestion
julia> design = SurveyDesign(apiclus1); 
```

%     [1, 1, 2  ...  2]
% popsize: [244586.316, 43527.8366, 36124.9061  ...  19331.022]
% sampsize: [3, 3, 3  ...  3]
% weights: [81528.772, 14509.2789, 12041.6354  ...  6443.674]
% allprobs: [0.0, 0.0001, 0.0001  ...  0.0002]
% \end{lstlisting}
\section{Estimation}

Survey.jl provides a range of estimators for survey data analysis. These include univariate statistics such as mean, median, total, and quantiles, as well as multivariate statistics such as regressions and ratios. For example, to estimate the mean of the \verb|:api99| column in the \verb|design| SurveyDesign:

\begin{lstlisting}
julia> mean(:api99, design)
1x1 DataFrame
    Row | mean    
        | Float64 
--------|--------
      1 | 606.978
    \end{lstlisting}
This command estimates the mean of column \verb|:api99|.

For multivariate statistics such as regressions\footnote{Regressions are performed using GLM.jl. Instead of passing a DataFrame, a survey design is passed to the function, maintaining a familiar interface. This approach of using multiple dispatch is applied to all estimators imported from other packages, ensuring consistency and ease of use.}:

\begin{lstlisting}
julia> glm(@formula(y ~ x),

>             my_design, Normal(), IdentityLink())

Reviewer: A semicolon missing? Since we're not showing any output?

Author: > A semicolon missing? Since we're not showing any output?

```suggestion
            my_design, Normal(), IdentityLink()); 
```

%             \end{itemize}
%         \item Usage:
%             \begin{enumerate}
%                 \item Generate replicate weights using bootstrap, jackknife, BRR, etc. 
%                 \item Using each replicate weight, calculate the estimate. 
%                 \item Calculate the standard error using the new set of estimates. 
%             \end{enumerate}
%         \end{itemize}

\subsection{Bootstrapping}

In the bootstrap method, each replicate \( r \) involves selecting a simple random sample of \( n_h - 1 \) primary sampling units (PSUs) with replacement from the \( n_h \) sample PSUs in stratum \( h \). The adjusted weight \( w_i'(r) \) for observation \( i \) in replicate \( r \) is calculated as:

% For bootstrap replicate $r (r = 1, \dots, R)$, an SRS of $n_h - 1$ PSUs is selected with replacement from the $n_h$ sample PSUs in stratum $h$. $m_{hj}(r)$ represents the number of times PSU $j$ of stratum $h$ is selected in replicate $r$.

% The adjusted weight $w_i'(r)$ for observation $i$ in replicate $r$ is calculated as:

\begin{equation}
    w_i'(r) = w_i(r) \times \frac{n_h}{n_h - 1} \times m_{h}(r)
\end{equation}

> Where \( w_i(r) \) denotes the initial weight for observation \( i \) within replicate \( r \), \( n_h \) is the total number of observations in stratum \( h \), and \( m_{h}(r) \) is the number of PSUs in stratum \( h \) that are selected in replicate \( r \)\cite{Lohr}.

Reviewer: Some space between `\( r \)` and `\cite{Lohr}` will be nice! Initially I thought we're indexing the variable `r`!

Author: > Some space between `\( r \)` and `\cite{Lohr}` will be nice! Initially I thought we're indexing the variable `r`!

```suggestion
Where \( w_i(r) \) denotes the initial weight for observation \( i \) within replicate \( r \), \( n_h \) is the total number of observations in stratum \( h \), and \( m_{h}(r) \) is the number of PSUs in stratum \( h \) that are selected in replicate \( r \) \cite{Lohr}.
```

%% The paper limit is 2 pages. Try pushing some things to the footnote, as space is reserved for it. Don't change the template. 

\documentclass{juliacon}
\setcounter{page}{1}

\usepackage{amsmath}

\begin{document}

\input{header}

\maketitle

\begin{abstract}

> Estimating variances in survey data analysis is challenging due to the complex nature of survey designs. It is typically done through resampling methods like bootstrapping. The Survey.jl package leverages Julia to provide an efficient framework for these resampling techniques, facilitating faster survey data analysis.

Reviewer 2: ```suggestion
Estimating variances in survey data analysis is challenging due to the complex nature of survey designs. It is typically done through resampling methods like bootstrapping, which can be computationally intensive. The Survey.jl package leverages Julia to provide an efficient framework for these resampling techniques, facilitating faster survey data analysis.
```

\end{abstract}

\section{Introduction}

> The growing volume of survey datasets necessitate more efficient analysis methods, particularly for variance estimation in complex survey designs. Computationally demanding resampling techniques, like bootstrapping and jackknife, are required when dealing with stratification, clustering, and unequal weights. 

Reviewer 2: ```suggestion
The growing volume of survey datasets necessitates more efficient analysis methods, particularly for variance estimation in complex survey designs. Computationally demanding resampling techniques, such as bootstrapping and jackknife, are required when dealing with stratification, clustering, and unequal weights. 
```

\\

> Many software packages exist for survey analysis\footnote{A comprehensive list is provided by \cite{SummarySurveyAnalysis}}. Notable examples include the R survey package, SAS/STAT, SPSS Complex Samples, Stata, and SUDAAN. The R survey package by Thomas Lumley\cite{lumley2004analysis} is widely recognized for its comprehensive capabilities and open-source availability. However, it lacks computational efficiency needed for large-scale data. Survey.jl leverages Julia to offer a faster resampling framework for variance estimation and survey data analysis.

Reviewer 2: ```suggestion
Many software packages exist for survey analysis\footnote{A comprehensive list is provided by \cite{SummarySurveyAnalysis}}. Notable examples include the R survey package, SAS/STAT, SPSS Complex Samples, Stata, and SUDAAN. The R survey package by Thomas Lumley\cite{lumley2004analysis} is widely recognized for its comprehensive capabilities and open-source availability. However, it is limited by R's computational efficiency, especially for large-scale data. Survey.jl leverages Julia to offer a faster resampling framework for variance estimation and survey data analysis.
```

% \href{https://www.hcp.med.harvard.edu/statistics/survey-soft/bascula.html}{Bascula}, 
% \href{https://www.hcp.med.harvard.edu/statistics/survey-soft/cenvar.html}{CENVAR}, 
% \href{https://www.hcp.med.harvard.edu/statistics/survey-soft/clusters.html}{CLUSTERS},  
% \href{https://www.cdc.gov/epiinfo/index.html}{Epi Info},  
% \href{https://www.statcan.gc.ca/eng/survey/methodology/Generalized_Estimation_System-eng.htm}{Generalized Estimation System (GES)},  
% \href{https://isr.umich.edu/}{IVEware},  
% \href{https://catalog.iastate.edu/azcourses/stat/}{PCCARP},  
% \href{https://cran.r-project.org/package=survey}{R survey package}, 
% \href{https://www.sas.com/en_us/home.html}{SAS/STAT},  
% \href{https://www.ibm.com/products/spss-statistics}{SPSS Complex Samples},  
% \href{https://www.stata.com/}{Stata},  
% \href{https://sudaanorder.rti.org/}{SUDAAN},  
% \href{https://www.census.gov/data/software/vplx.html}{VPLX},  
% \href{https://www.westat.com/wesvar/}{WesVar}

% The survey package in R by Thomas Lumely \cite{lumley2004analysis} is the widely used open-source package. 

\section{Survey design}

> A \verb|SurveyDesign| object can be created to incorporate the sampling design. This object requires the following parameters: \verb|data::DataFrame|, which is the survey data in the form of a \verb|DataFrame|; \verb|clusters::Symbol|, specifying the column name containing the clusters; \verb|strata::Symbol|, specifying the column name containing the strata; \verb|weights::Symbol|, indicating the column name containing the sampling weights; and \verb|popsize::Symbol|, indicating the column name containing the population size.\footnote{Internaly, there is a single constructor for all types of surveys. Every survey is assumed to be a complex survey. If there is no stratification, we assume that everything is part of one stratum. If there is no clustering, we assume each member is a cluster.}  

Reviewer 2: ```suggestion
The \verb|SurveyDesign| object instantiates a sampling design. This object requires the following parameters: \verb|data::DataFrame|, which is the survey data in the form of a \verb|DataFrame|; \verb|clusters::Symbol|, specifying the column name containing the clusters; \verb|strata::Symbol|, specifying the column name containing the strata; \verb|weights::Symbol|, indicating the column name containing the sampling weights; and \verb|popsize::Symbol|, indicating the column name containing the population size.\footnote{Internaly, there is a single constructor for all types of surveys. Every survey is assumed to be a complex survey. If there is no stratification, we assume that everything is part of one stratum. If there is no clustering, we assume each member is a cluster.}  
```

                    weights=:WTMEC2YR);
    \end{lstlisting}
Consider another example, a cluster sample based on the Academic Performance Index for all California schools based on standardised testing of students. There is no stratification in this example. 

\begin{lstlisting}
julia> apiclus1 = load_data("apiclus1"); 
# CSV dataframe included with the package

design = SurveyDesign(apiclus1; 
            clusters=:dnum, weights=:pw);
    \end{lstlisting}
% \begin{lstlisting}
% julia> nhanes = load_data("nhanes") 
% # CSV dataframe included with the package

% julia> SurveyDesign(nhanes; clusters=:SDMVPSU,
%                     strata=:SDMVSTRA, 
%                     weights=:WTMEC2YR)

% SurveyDesign:
% data: 8591 x 11 DataFrame
% strata: SDMVSTRA
%     [83, 84, 86  ...  81]
% cluster: SDMVPSU
%     [1, 1, 2  ...  2]
% popsize: [244586.316, 43527.8366, 36124.9061  ...  19331.022]
% sampsize: [3, 3, 3  ...  3]
% weights: [81528.772, 14509.2789, 12041.6354  ...  6443.674]
% allprobs: [0.0, 0.0001, 0.0001  ...  0.0002]
% \end{lstlisting}
\section{Estimation}

Survey.jl provides basic univariate and multivariate estimators for efficient survey data analysis.

> For univariate statistics such as mean, median, total, and quantiles, the following examples illustrate their usage:

Reviewer 2: ```suggestion
Survey.jl provides a range of estimators for survey data analysis. These include univariate statistics such as mean, median, total, and quantiles, as well as multivariate statistics such as regressions and ratios. For example, to estimate the mean of the \verb|:api99| column in the \verb|design| SurveyDesign:
```

\begin{lstlisting}

> julia> mean(:api99, design)

Reviewer 2: ```suggestion
julia> mean(:api99, my_design)
```

1x1 DataFrame
    Row | mean    
        | Float64 
--------|--------
      1 | 606.978
    \end{lstlisting}
This command estimates the mean of column \verb|:api99|.

> For multivariate statistics such as regressions\footnote{Regressions are performed using GLM.jl. Instead of passing a DataFrame, a survey design is passed to the function, maintaining a familiar interface. This approach of using multiple dispatch is applied to all estimators imported from other packages, ensuring consistency and ease of use.}:

Reviewer 2: ```suggestion
Regression models\footnote{Regressions are performed using GLM.jl. Instead of passing a DataFrame, a survey design is passed to the function, maintaining a familiar interface. This approach of using multiple dispatch is applied to all estimators imported from other packages, ensuring consistency and ease of use.} can be fitted directly to \verb|SurveyDesign| objects:
```

\begin{lstlisting}

> julia> glm(@formula(y ~ x),

Reviewer 2: ```suggestion
julia> glm(@formula(:api99 ~ :api00),
```

            my_design, Normal(), IdentityLink())
\end{lstlisting}

% And ratio: 

% \begin{lstlisting}
% julia> ratio(:y, :x, my_design)
% \end{lstlisting}

\section{Replicate weights}

> The standard error of an estimator measures the average amount of variability or uncertainty in the estimated value. Standard errors are often provided alongside point estimates in various statistical packages.
> 
> To estimate standard errors for complex survey designs, Survey.jl uses replicate weights, generated through resampling techniques such as bootstrap and jackknife. Each replicate sample represents a plausible variation of the original sample, allowing for the estimation of variability as if the sampling were repeated multiple times.

Reviewer 2: ```suggestion
Survey data analysis requires variance estimation to quantify the uncertainty in the estimates. This is particularly important for complex survey designs, where the standard error of an estimator is not directly available. 

Survey.jl provides a framework for estimating standard errors by generating replicate weights through resampling techniques such as bootstrapping and jackknife. Each replicate sample represents a plausible variation of the original sample, allowing for the estimation of variability as if the sampling were repeated multiple times.
```

> The estimate is calculated for each replicate, and then the standard error is computed from the distribution of these estimates. 

Reviewer 2: ```suggestion
An estimate of the variance is obtained by calculating the variance of the estimates across the replicates.
```

%                 \item Calculate the standard error using the new set of estimates. 
%             \end{enumerate}
%         \end{itemize}

\subsection{Bootstrapping}

In the bootstrap method, each replicate \( r \) involves selecting a simple random sample of \( n_h - 1 \) primary sampling units (PSUs) with replacement from the \( n_h \) sample PSUs in stratum \( h \). The adjusted weight \( w_i'(r) \) for observation \( i \) in replicate \( r \) is calculated as:

% For bootstrap replicate $r (r = 1, \dots, R)$, an SRS of $n_h - 1$ PSUs is selected with replacement from the $n_h$ sample PSUs in stratum $h$. $m_{hj}(r)$ represents the number of times PSU $j$ of stratum $h$ is selected in replicate $r$.

% The adjusted weight $w_i'(r)$ for observation $i$ in replicate $r$ is calculated as:

\begin{equation}
    w_i'(r) = w_i(r) \times \frac{n_h}{n_h - 1} \times m_{h}(r)
\end{equation}

Where \( w_i(r) \) denotes the initial weight for observation \( i \) within replicate \( r \), \( n_h \) is the total number of observations in stratum \( h \), and \( m_{h}(r) \) is the number of PSUs in stratum \( h \) that are selected in replicate \( r \)\cite{Lohr}.

> \verb|bootweights| can be used to generate \verb|ReplicateDesign{BootstrapReplicates}| from a \verb|SurveyDesign|. 

Reviewer 2: ```suggestion
Survey.jl supports bootstrapping for variance estimation using the \verb|bootweights| function. This function generates replicate weights for bootstrapping from a \verb|SurveyDesign| object, and the resulting object is a \verb|ReplicateDesign{BootstrapReplicates}|.
```

\begin{lstlisting}

> julia> bdesign = bootweights(design; replicates = 1000)

Reviewer 2: ```suggestion
julia> my_bootstrapped_design = bootweights(my_design; replicates = 1000)
```

\end{lstlisting}

% \begin{lstlisting}
% julia> srs = SurveyDesign(apisrs; weights=:pw);

% julia> bsrs = bootweights(srs; replicates = 1000)
% ReplicateDesign{BootstrapReplicates}:
% data: 200x1045 DataFrame
% strata: none
% cluster: none
% popsize: [6194.0, 6194.0, 6194.0  ...  6194.0]
% sampsize: [200, 200, 200  ...  200]
% weights: [30.97, 30.97, 30.97  ...  30.97]
% allprobs: [0.0323, 0.0323, 0.0323  ...  0.0323]
% type: bootstrap
% replicates: 1000
% \end{lstlisting}

> The replicate design object facilitates variance estimation. When a function receives a \verb|ReplicateDesign| rather than a \verb|SurveyDesign|, it provides the standard error along with the point estimate.

Reviewer 2: ```suggestion
\verb|ReplicateDesign| objects can be passed to summary statistics functions, which will return the standard error of the estimator alongside the point estimate:
```

For example: 
\begin{lstlisting}

> julia> mean(:api99, bdesign)

Reviewer 2: ```suggestion
julia> mean(:api99, my_bootstrapped_design)
```

1x2 DataFrame
    Row | mean     SE      
        | Float64  Float64 
--------|-----------------
      1 | 606.978  24.7505
    \end{lstlisting}

> For each replicate $r$, $\hat{\theta}^*_r$ is the estimator of $\theta$, calculated the same way as $\hat{\theta}$ but using weights $w_i'(r)$ instead of the original weights $w_i$. The variance of the estimator is given by: 

Reviewer 2: ```suggestion
For each replicate $r$, $\hat{\theta}^*_r$ is calculated using the adjusted weights $w_i'(r)$. $\hat{\theta}^*$ instead of the original weights $w_i$. The variance of the estimator is estimated as: 
```

\begin{equation}
        \hat{V}_B(\hat{\theta}) = \dfrac{1}{R-1}\sum_{r=1}^{R} (\hat{\theta}^*_r - \hat{\theta})^2.
\end{equation}

\subsection{Jackknife}

> In the jackknife method, each PSU is systematically omitted one at a time to create replicates. The adjusted weight \( w_{i(hj)} \) for observation \( i \) when PSU \( j \) in stratum \( h \) is omitted is calculated as:

Reviewer 2: ```suggestion
Survey.jl also supports the jackknife method for variance estimation. The package implements the leave-one-out jackknife, which omits each PSU one at a time to create replicates. The adjusted weight \( w_{i(hj)} \) for observation \( i \) when PSU \( j \) in stratum \( h \) is omitted is calculated as:
```

\begin{equation}
    w_{i(hj)} = \begin{cases}
        w_i & i \notin h\\
    0 & i \in j_{h} \\
    \dfrac{n_h}{n_h - 1} w_i &  i \in h \text{ and } i \notin j_{h}
    \end{cases} %% Fix equation
    \end{equation} \cite{Lohr}

> \verb|jackknifeweights| can be used to generate \verb|ReplicateDesign{JackknifeReplicates}| from a \verb|SurveyDesign|. 

Reviewer 2: ```suggestion
The \verb|jackknifeweights| function takes in a \verb|SurveyDesign| object and returns a \verb|ReplicateDesign{JackknifeReplicates}| object:
```

\begin{lstlisting}

>     julia> design = jackknifeweights(design)

Reviewer 2: ```suggestion
    julia> my_jackknife_design = jackknifeweights(my_design)
```

    \end{lstlisting}

% \begin{lstlisting}
% julia> jsrs = jackknifeweights(srs)
% ReplicateDesign{JackknifeReplicates}:
% data: 200x245 DataFrame
% strata: none
% cluster: none
% popsize: [6194.0, 6194.0, 6194.0  ...  6194.0]
% sampsize: [200, 200, 200  ...  200]
% weights: [30.97, 30.97, 30.97  ...  30.97]
% allprobs: [0.0323, 0.0323, 0.0323  ...  0.0323]
% type: jackknife
% replicates: 200
% \end{lstlisting}

> This object can be passed to estimators to obtain an estimate of variance alongside the point estimate. 

Reviewer 2: ```suggestion
This \verb|ReplicateDesign| object can also be passed to summary statistics functions to obtain an estimate of the variance of the estimator alongside its point estimate.
```

> $\hat{\theta}$ represents the estimator computed using the original weights, and $\hat{\theta_{(hj)}}$ represents the estimator computed from the replicate weights obtained when PSU $j$ from cluster $h$ is removed. The variance is given by: 

Reviewer 2: ```suggestion
$\hat{\theta}$ represents the estimator computed using the original weights, and $\hat{\theta_{(hj)}}$ represents the estimator computed from the replicate weights obtained when PSU $j$ from cluster $h$ is removed. The variance is estimated as: 
```

\begin{equation}
\hat{V}_{\text{JK}}(\hat{\theta}) = \sum_{h = 1}^H \dfrac{n_h - 1}{n_h}\sum_{j = 1}^{n_h}(\hat{\theta}_{(hj)} - \hat{\theta})^2
\end{equation}

\subsection{Extending variance estimation}

> Survey.jl supports variance estimation for basic estimators such as mean, quantile, ratio, and regressions. The package is designed to extend this capability to any estimator through a generalized approach.
> 
> The \verb|Survey.variance| function can be applied to replicate designs to estimate the variance of any estimator function, allowing users and developers to extend variance estimation to custom estimators.

Reviewer 2: ```suggestion
Survey.jl currently supports variance estimation for the summary statistics functions provided by the package, but the framework can be extended to custom estimators. The \verb|variance| function can be applied to \verb|ReplicateDesign| objects to estimate the variance of an estimator function, such as \verb|Survey.mean|.
```

% \begin{lstlisting}
% function variance(
%     design::ReplicateDesign,
%     func::Function, ...)
% \end{lstlisting}

% This flexibility allows users and developers to extend variance estimation to custom estimators.

% at appropriate place in your \TeX{} file or in bibliography file.

\section{Conclusions}

> Survey.jl offers an efficient framework for survey data analysis. Its functionality has been tested against R's survey package, and future development aims to port all features from R.

Reviewer 2: ```suggestion
Survey.jl provides a comprehensive framework for survey data analysis, leveraging Julia's computational efficiency. The package has been tested against R's survey package, and future development aims to port all features from R.
```
